integration_tests/test_gravity.py

```diff
@@ -396,8 +394,106 @@ def test_direct_token_mapping(gravity):
+        geth = gravity.geth
+        cli = gravity.cronos.cosmos_cli()
+        cronos_w3 = gravity.cronos.w3
+
+        # deploy test erc20 contract
+        erc20 = deploy_contract(
+            geth,
+            CONTRACTS["TestERC20A"],
+        )
+
+        # deploy gravity cancellation contract
+        cancel_contract = deploy_contract(
+            cronos_w3,
+            CONTRACTS["CronosGravityCancellation"],
+        )
+
+        balance = erc20.caller.balanceOf(ADDRS["validator"])
+        assert balance == 100000000000000000000000000
+        amount = 1000
+
+        print("send to cronos crc21")
+        recipient = HexBytes(ADDRS["community"])
+        send_to_cosmos(gravity.contract, erc20, recipient, amount, KEYS["validator"])
+        assert erc20.caller.balanceOf(ADDRS["validator"]) == balance - amount
+
+        denom = f"gravity{erc20.address}"
+
+        crc21_contract = None
+
+        def check_auto_deployment():
+            "check crc21 contract auto deployed, and the crc21 balance"
+            nonlocal crc21_contract
+            try:
+                rsp = cli.query_contract_by_denom(denom)
+            except AssertionError:
+                # not deployed yet
+                return False
+            assert len(rsp["auto_contract"]) > 0
+            crc21_contract = cronos_w3.eth.contract(
+                address=rsp["auto_contract"], abi=cronos_crc21_abi()
+            )
+            return crc21_contract.caller.balanceOf(recipient) == amount
+
+        def get_id_from_receipt(receipt):
+            "check the id after sendToChain call"
+            for _, log in enumerate(receipt.logs):
+                if log.topics[0] == HexBytes(
+                    abi.event_signature_to_log_topic(
+                        "__CronosSendToChainResponse(uint256)"
+                    )
+                ):
+                    return log.data
+            return "0x0000000000000000000000000000000000000000000000000000000000000000"
+
+        wait_for_fn("send-to-crc20", check_auto_deployment)
+
+        def check_fund():
+            v = crc21_contract.caller.balanceOf(ADDRS["community"])
+            return v == amount
+
+        wait_for_fn("send-to-ethereum", check_fund)
+
+        # send it back to erc20
+        tx = crc21_contract.functions.send_to_chain(
+            ADDRS["validator"], amount, 0, 1
+        ).buildTransaction({"from": ADDRS["community"]})
+        txreceipt = send_transaction(cronos_w3, tx, KEYS["community"])
+        # CRC20 emit 3 logs for send_to_chain:
+        # burn
+        # __CronosSendToChain
+        # __CronosSendToChainResponse
+        assert len(txreceipt.logs) == 3
+        tx_id = get_id_from_receipt(txreceipt)
+        assert txreceipt.status == 1, "should success"
+
+        def check_deduction():
+            v = crc21_contract.caller.balanceOf(ADDRS["community"])
+            return v == 0
+
+        wait_for_fn("check deduction", check_deduction)
```

Comment:
should be able to check balance immediately, no need to wait?

Reply:
yes, changed 👌